File: examples/analytics-api/python/example_dashboard.py
"""
Complete Dashboard Example for TESAIoT Analytics API

This example demonstrates building a comprehensive IoT analytics
dashboard that combines all four analytics domains:
- Anomaly Detection
- Pattern Recognition
- AI Insights
- Connectivity Metrics

Author: TESAIoT Team
License: Apache 2.0
"""

import asyncio
import os
import json
from datetime import datetime
from dataclasses import dataclass
from typing import Dict, Any, List, Optional
from analytics_client import AnalyticsClient, TimeRange


@dataclass
class DashboardData:
    """Container for all dashboard data"""
    timestamp: str
    fleet_health: Dict[str, Any]
    anomalies: Dict[str, Any]
    clusters: Dict[str, Any]
    insights: List[Dict[str, Any]]
    connectivity: Dict[str, Any]
    alerts: List[Dict[str, Any]]


async def collect_dashboard_data() -> DashboardData:
    """Collect all data needed for the dashboard"""
    async with AnalyticsClient() as client:
        # Collect all data in parallel for efficiency
        anomalies_task = client.get_anomalies(limit=100, severity_filter=["critical", "high", "medium"])
        timeline_task = client.get_anomaly_timeline(days=30)
        clusters_task = client.get_clusters(metric_name="temperature", n_clusters=5)
        insights_task = client.get_insights(days=7)
        connectivity_task = client.get_connectivity_status()
        latency_task = client.get_latency_stats(hours=24)
        throughput_task = client.get_throughput_stats(hours=24)
        quality_task = client.get_connection_quality()

        # Await all tasks
        anomalies, timeline, clusters, insights, connectivity, latency, throughput, quality = await asyncio.gather(
            anomalies_task, timeline_task, clusters_task, insights_task,
            connectivity_task, latency_task, throughput_task, quality_task
        )

        # Process and structure the data
        alerts = generate_alerts(anomalies, connectivity, latency, quality)

        fleet_health = calculate_fleet_health(
            anomalies, connectivity, insights, latency
        )

        return DashboardData(
            timestamp=datetime.utcnow().isoformat() + "Z",
            fleet_health=fleet_health,
            anomalies={
                "summary": anomalies.get("summary", {}),
                "recent": anomalies.get("anomalies", [])[:10],
                "timeline": timeline.get("timeline", []),
                "trend": timeline.get("trend", {})
            },
            clusters={
                "clusters": clusters.get("clusters", []),
                "silhouette_score": clusters.get("silhouette_score", 0),
                "outliers": clusters.get("outliers", [])[:5]
            },
            insights=insights.get("insights", []),
            connectivity={
                "status": connectivity.get("summary", {}),
                "devices": connectivity.get("devices", [])[:20],
                "latency": latency.get("summary", {}),
                "throughput": throughput.get("summary", {}),
                "quality": quality.get("summary", {})
            },
            alerts=alerts
        )


def calculate_fleet_health(anomalies, connectivity, insights, latency) -> Dict[str, Any]:
    """Calculate overall fleet health score"""
    scores = {}

    # Anomaly score (0-100, lower anomaly rate = higher score)
    anomaly_summary = anomalies.get("summary", {})
    total_devices = connectivity.get("summary", {}).get("total", 100)
    anomaly_count = anomaly_summary.get("total", 0)
    anomaly_rate = anomaly_count / max(total_devices, 1)
    scores["anomaly"] = max(0, 100 - (anomaly_rate * 1000))

    # Connectivity score (percentage online)
    conn_summary = connectivity.get("summary", {})
    online = conn_summary.get("online", 0)
    total = conn_summary.get("total", 1)
    scores["connectivity"] = (online / max(total, 1)) * 100

    # Latency score (0-100, lower latency = higher score)
    p95_latency = latency.get("summary", {}).get("p95_latency_ms", 0)
    scores["latency"] = max(0, 100 - (p95_latency / 10))

    # Insights severity score
    insights_list = insights.get("insights", [])
    critical_count = sum(1 for i in insights_list if i.get("severity") == "critical")
    warning_count = sum(1 for i in insights_list if i.get("severity") == "warning")
    scores["insights"] = max(0, 100 - (critical_count * 20) - (warning_count * 5))

    # Calculate weighted overall score
    overall = (
        scores["anomaly"] * 0.3 +
        scores["connectivity"] * 0.3 +
        scores["latency"] * 0.2 +
        scores["insights"] * 0.2
    )

    return {
        "overall_score": round(overall, 1),
        "component_scores": scores,
        "status": get_health_status(overall)
    }


def get_health_status(score: float) -> str:
    """Convert health score to status label"""
    if score >= 90:
        return "EXCELLENT"
    elif score >= 70:
        return "GOOD"
    elif score >= 50:
        return "FAIR"
    elif score >= 30:
        return "POOR"
    else:
        return "CRITICAL"


def generate_alerts(anomalies, connectivity, latency, quality) -> List[Dict[str, Any]]:
    """Generate alerts based on current data"""
    alerts = []

    # Critical anomaly alerts
    anomaly_summary = anomalies.get("summary", {})
    critical_count = anomaly_summary.get("by_severity", {}).get("critical", 0)
    if critical_count > 0:
        alerts.append({
            "level": "critical",
            "type": "anomaly",
            "title": f"{critical_count} Critical Anomalies Detected",
            "description": "Immediate investigation recommended"
        })

    # Offline device alerts
    conn_summary = connectivity.get("summary", {})
    offline = conn_summary.get("offline", 0)
    total = conn_summary.get("total", 1)
    offline_pct = (offline / max(total, 1)) * 100

    if offline_pct > 20:
        alerts.append({
            "level": "critical",
            "type": "connectivity",
            "title": f"{offline} Devices Offline ({offline_pct:.0f}%)",
            "description": "Network connectivity issue detected"
        })
    elif offline > 0:
        alerts.append({
            "level": "warning",
            "type": "connectivity",
            "title": f"{offline} Device(s) Offline",
            "description": "Some devices are not responding"
        })

    # Latency alerts
    p95 = latency.get("summary", {}).get("p95_latency_ms", 0)
    if p95 > 1000:
        alerts.append({
            "level": "critical",
            "type": "latency",
            "title": f"High Latency: {p95:.0f}ms P95",
            "description": "Network performance severely degraded"
        })
    elif p95 > 500:
        alerts.append({
            "level": "warning",
            "type": "latency",
            "title": f"Elevated Latency: {p95:.0f}ms P95",
            "description": "Network performance degraded"
        })

    # Quality alerts
    poor_devices = quality.get("summary", {}).get("distribution", {}).get("poor", 0)
    if poor_devices > 5:
        alerts.append({
            "level": "warning",
            "type": "quality",
            "title": f"{poor_devices} Devices with Poor Connection Quality",
            "description": "Review device connections and network path"
        })

    return alerts


def render_dashboard(data: DashboardData):
    """Render the dashboard to console"""
    print("\n" + "=" * 80)
    print(" " * 25 + "TESAIoT ANALYTICS DASHBOARD")
    print("=" * 80)
    print(f"Generated: {data.timestamp}")

    # Fleet Health Overview
    print("\n" + "-" * 80)
    print(" FLEET HEALTH")
    print("-" * 80)

    health = data.fleet_health
    score = health["overall_score"]
    status = health["status"]

    # Visual health bar
    bar_len = int(score / 5)
    bar = "" * bar_len + "" * (20 - bar_len)
    print(f"\n  Overall: [{bar}] {score:.0f}/100 ({status})")

    print(f"\n  Component Scores:")
    for component, comp_score in health["component_scores"].items():
        print(f"    {component.capitalize():15} {comp_score:.0f}/100")

    # Alerts Section
    if data.alerts:
        print("\n" + "-" * 80)
        print(" ACTIVE ALERTS")
        print("-" * 80)

        for alert in data.alerts:
            icon = "!" if alert["level"] == "critical" else "!" if alert["level"] == "warning" else "i"
            print(f"\n  [{icon}] {alert['title']}")
            print(f"      {alert['description']}")

    # Anomaly Detection Section
    print("\n" + "-" * 80)
    print(" ANOMALY DETECTION")
    print("-" * 80)

    anom = data.anomalies
    summary = anom["summary"]
    trend = anom.get("trend", {})

    print(f"\n  Total Anomalies (7 days): {summary.get('total', 0)}")
    print(f"  By Severity:")
    for sev, count in summary.get("by_severity", {}).items():
        print(f"    {sev.capitalize():10} {count}")

    trend_dir = trend.get("direction", "stable")
    trend_pct = trend.get("percent_change", 0)
    trend_icon = "" if trend_dir == "decreasing" else "" if trend_dir == "increasing" else ""
    print(f"\n  Trend: {trend_icon} {trend_dir} ({trend_pct:+.1f}%)")

    # Pattern Recognition Section
    print("\n" + "-" * 80)
    print(" PATTERN RECOGNITION")
    print("-" * 80)

    clusters = data.clusters
    print(f"\n  Clusters: {len(clusters['clusters'])}")
    print(f"  Silhouette Score: {clusters['silhouette_score']:.3f}")

    for cluster in clusters["clusters"]:
        print(f"\n    Cluster {cluster['cluster_id']}: {cluster['device_count']} devices")
        chars = cluster.get("characteristics", {})
        if chars:
            print(f"      Mean: {chars.get('mean', 0):.2f}, StdDev: {chars.get('std_dev', 0):.2f}")

    if clusters["outliers"]:
        print(f"\n  Outliers: {len(clusters['outliers'])} devices")

    # AI Insights Section
    print("\n" + "-" * 80)
    print(" AI INSIGHTS")
    print("-" * 80)

    insights = data.insights
    print(f"\n  Total Insights: {len(insights)}")

    critical_insights = [i for i in insights if i.get("severity") == "critical"]
    if critical_insights:
        print(f"\n  Critical Insights:")
        for i in critical_insights[:3]:
            print(f"    ! {i['title']}")

    actionable = [i for i in insights if i.get("actionable")]
    if actionable:
        print(f"\n  Actionable Recommendations ({len(actionable)}):")
        for i in actionable[:3]:
            print(f"    - {i['title']}")

    # Connectivity Section
    print("\n" + "-" * 80)
    print(" CONNECTIVITY")
    print("-" * 80)

    conn = data.connectivity
    status = conn["status"]

    print(f"\n  Devices: {status.get('online', 0)}/{status.get('total', 0)} online")

    latency = conn["latency"]
    print(f"\n  Latency:")
    print(f"    Average: {latency.get('avg_latency_ms', 0):.1f} ms")
    print(f"    P95: {latency.get('p95_latency_ms', 0):.1f} ms")

    throughput = conn["throughput"]
    print(f"\n  Throughput (24h):")
    print(f"    Total Messages: {throughput.get('total_messages', 0):,}")
    print(f"    Avg/Hour: {throughput.get('avg_per_hour', 0):,.0f}")

    quality = conn["quality"]
    print(f"\n  Connection Quality: {quality.get('average_quality_score', 0):.0f}/100")

    # Footer
    print("\n" + "=" * 80)
    print(" " * 30 + "END OF REPORT")
    print("=" * 80)


async def dashboard_refresh_loop(interval_seconds: int = 60, max_iterations: int = 5):
    """Continuously refresh dashboard"""
    print("Starting dashboard refresh loop...")
    print(f"Refresh interval: {interval_seconds} seconds")
    print("Press Ctrl+C to stop\n")

    iteration = 0
    while iteration < max_iterations:
        iteration += 1

        try:
            # Collect fresh data
            data = await collect_dashboard_data()

            # Clear screen (simulated with newlines in this demo)
            print("\n" * 2)

            # Render dashboard
            render_dashboard(data)

            print(f"\n[Refresh {iteration}/{max_iterations} - "
                  f"Next refresh in {interval_seconds} seconds]")

            # Wait for next refresh
            await asyncio.sleep(interval_seconds)

        except KeyboardInterrupt:
            print("\n\nDashboard stopped by user.")
            break
        except Exception as e:
            print(f"\nError refreshing dashboard: {e}")
            await asyncio.sleep(10)

    print("\nDashboard loop completed.")


async def export_dashboard_json(output_file: str = "dashboard_data.json"):
    """Export dashboard data to JSON file"""
    print(f"Exporting dashboard data to {output_file}...")

    data = await collect_dashboard_data()

    # Convert dataclass to dict for JSON serialization
    export_data = {
        "timestamp": data.timestamp,
        "fleet_health": data.fleet_health,
        "anomalies": data.anomalies,
        "clusters": data.clusters,
        "insights": data.insights,
        "connectivity": data.connectivity,
        "alerts": data.alerts
    }

    with open(output_file, "w") as f:
        json.dump(export_data, f, indent=2)

    print(f"Dashboard data exported to {output_file}")
    print(f"File size: {os.path.getsize(output_file)} bytes")


async def main():
    """Main dashboard entry point"""
    import sys

    if len(sys.argv) > 1:
        if sys.argv[1] == "--loop":
            interval = int(sys.argv[2]) if len(sys.argv) > 2 else 60
            await dashboard_refresh_loop(interval_seconds=interval)
        elif sys.argv[1] == "--export":
            output = sys.argv[2] if len(sys.argv) > 2 else "dashboard_data.json"
            await export_dashboard_json(output)
        elif sys.argv[1] == "--help":
            print("TESAIoT Analytics Dashboard")
            print("\nUsage:")
            print("  python example_dashboard.py          # Single refresh")
            print("  python example_dashboard.py --loop   # Continuous refresh (60s)")
            print("  python example_dashboard.py --loop 30  # Custom interval")
            print("  python example_dashboard.py --export  # Export to JSON")
            return
    else:
        # Single dashboard render
        print("Collecting dashboard data...")
        data = await collect_dashboard_data()
        render_dashboard(data)


if __name__ == "__main__":
    asyncio.run(main())

File: examples/analytics-api/javascript/example-dashboard.js
/**
 * Complete Dashboard Example for TESAIoT Analytics API (JavaScript)
 *
 * This example demonstrates building a comprehensive IoT analytics
 * dashboard that combines all four analytics domains.
 *
 * Author: TESAIoT Team
 * License: Apache 2.0
 */

import { AnalyticsClient, TimeRange } from './analytics-client.js';

/**
 * Collect all dashboard data in parallel
 */
async function collectDashboardData(client) {
    const [
        anomalies,
        timeline,
        clusters,
        insights,
        connectivity,
        latency,
        throughput,
        quality
    ] = await Promise.all([
        client.getAnomalies({ limit: 100, severityFilter: ['critical', 'high', 'medium'] }),
        client.getAnomalyTimeline({ days: 30 }),
        client.getClusters({ metricName: 'temperature', nClusters: 5 }),
        client.getInsights({ days: 7 }),
        client.getConnectivityStatus(),
        client.getLatencyStats({ hours: 24 }),
        client.getThroughputStats({ hours: 24 }),
        client.getConnectionQuality()
    ]);

    const alerts = generateAlerts(anomalies, connectivity, latency, quality);
    const fleetHealth = calculateFleetHealth(anomalies, connectivity, insights, latency);

    return {
        timestamp: new Date().toISOString(),
        fleetHealth,
        anomalies: {
            summary: anomalies.summary || {},
            recent: (anomalies.anomalies || []).slice(0, 10),
            timeline: timeline.timeline || [],
            trend: timeline.trend || {}
        },
        clusters: {
            clusters: clusters.clusters || [],
            silhouetteScore: clusters.silhouette_score || 0,
            outliers: (clusters.outliers || []).slice(0, 5)
        },
        insights: insights.insights || [],
        connectivity: {
            status: connectivity.summary || {},
            devices: (connectivity.devices || []).slice(0, 20),
            latency: latency.summary || {},
            throughput: throughput.summary || {},
            quality: quality.summary || {}
        },
        alerts
    };
}

/**
 * Calculate overall fleet health score
 */
function calculateFleetHealth(anomalies, connectivity, insights, latency) {
    const scores = {};

    // Anomaly score
    const anomalySummary = anomalies.summary || {};
    const totalDevices = connectivity.summary?.total || 100;
    const anomalyCount = anomalySummary.total || 0;
    const anomalyRate = anomalyCount / Math.max(totalDevices, 1);
    scores.anomaly = Math.max(0, 100 - (anomalyRate * 1000));

    // Connectivity score
    const connSummary = connectivity.summary || {};
    const online = connSummary.online || 0;
    const total = connSummary.total || 1;
    scores.connectivity = (online / Math.max(total, 1)) * 100;

    // Latency score
    const p95Latency = latency.summary?.p95_latency_ms || 0;
    scores.latency = Math.max(0, 100 - (p95Latency / 10));

    // Insights severity score
    const insightsList = insights.insights || [];
    const criticalCount = insightsList.filter(i => i.severity === 'critical').length;
    const warningCount = insightsList.filter(i => i.severity === 'warning').length;
    scores.insights = Math.max(0, 100 - (criticalCount * 20) - (warningCount * 5));

    // Calculate weighted overall score
    const overall =
        scores.anomaly * 0.3 +
        scores.connectivity * 0.3 +
        scores.latency * 0.2 +
        scores.insights * 0.2;

    return {
        overallScore: Math.round(overall * 10) / 10,
        componentScores: scores,
        status: getHealthStatus(overall)
    };
}

/**
 * Convert health score to status label
 */
function getHealthStatus(score) {
    if (score >= 90) return 'EXCELLENT';
    if (score >= 70) return 'GOOD';
    if (score >= 50) return 'FAIR';
    if (score >= 30) return 'POOR';
    return 'CRITICAL';
}

/**
 * Generate alerts based on current data
 */
function generateAlerts(anomalies, connectivity, latency, quality) {
    const alerts = [];

    // Critical anomaly alerts
    const criticalCount = anomalies.summary?.by_severity?.critical || 0;
    if (criticalCount > 0) {
        alerts.push({
            level: 'critical',
            type: 'anomaly',
            title: `${criticalCount} Critical Anomalies Detected`,
            description: 'Immediate investigation recommended'
        });
    }

    // Offline device alerts
    const offline = connectivity.summary?.offline || 0;
    const total = connectivity.summary?.total || 1;
    const offlinePct = (offline / Math.max(total, 1)) * 100;

    if (offlinePct > 20) {
        alerts.push({
            level: 'critical',
            type: 'connectivity',
            title: `${offline} Devices Offline (${offlinePct.toFixed(0)}%)`,
            description: 'Network connectivity issue detected'
        });
    } else if (offline > 0) {
        alerts.push({
            level: 'warning',
            type: 'connectivity',
            title: `${offline} Device(s) Offline`,
            description: 'Some devices are not responding'
        });
    }

    // Latency alerts
    const p95 = latency.summary?.p95_latency_ms || 0;
    if (p95 > 1000) {
        alerts.push({
            level: 'critical',
            type: 'latency',
            title: `High Latency: ${p95.toFixed(0)}ms P95`,
            description: 'Network performance severely degraded'
        });
    } else if (p95 > 500) {
        alerts.push({
            level: 'warning',
            type: 'latency',
            title: `Elevated Latency: ${p95.toFixed(0)}ms P95`,
            description: 'Network performance degraded'
        });
    }

    // Quality alerts
    const poorDevices = quality.summary?.distribution?.poor || 0;
    if (poorDevices > 5) {
        alerts.push({
            level: 'warning',
            type: 'quality',
            title: `${poorDevices} Devices with Poor Connection Quality`,
            description: 'Review device connections and network path'
        });
    }

    return alerts;
}

/**
 * Render dashboard to console
 */
function renderDashboard(data) {
    console.log('\n' + '='.repeat(80));
    console.log(' '.repeat(25) + 'TESAIoT ANALYTICS DASHBOARD');
    console.log('='.repeat(80));
    console.log(`Generated: ${data.timestamp}`);

    // Fleet Health
    console.log('\n' + '-'.repeat(80));
    console.log(' FLEET HEALTH');
    console.log('-'.repeat(80));

    const { overallScore, status, componentScores } = data.fleetHealth;
    const barLen = Math.floor(overallScore / 5);
    const bar = '\u2588'.repeat(barLen) + '\u2591'.repeat(20 - barLen);
    console.log(`\n  Overall: [${bar}] ${overallScore}/100 (${status})`);

    console.log('\n  Component Scores:');
    for (const [component, score] of Object.entries(componentScores)) {
        console.log(`    ${component.padEnd(15)} ${Math.round(score)}/100`);
    }

    // Alerts
    if (data.alerts.length > 0) {
        console.log('\n' + '-'.repeat(80));
        console.log(' ACTIVE ALERTS');
        console.log('-'.repeat(80));

        for (const alert of data.alerts) {
            const icon = alert.level === 'critical' ? '\u2757' : '\u26A0';
            console.log(`\n  [${icon}] ${alert.title}`);
            console.log(`      ${alert.description}`);
        }
    }

    // Anomalies
    console.log('\n' + '-'.repeat(80));
    console.log(' ANOMALY DETECTION');
    console.log('-'.repeat(80));

    const anomSummary = data.anomalies.summary;
    console.log(`\n  Total Anomalies (7 days): ${anomSummary.total || 0}`);

    if (anomSummary.by_severity) {
        console.log('  By Severity:');
        for (const [sev, count] of Object.entries(anomSummary.by_severity)) {
            console.log(`    ${sev.padEnd(10)} ${count}`);
        }
    }

    const trend = data.anomalies.trend;
    const trendIcon = trend.direction === 'decreasing' ? '\u2193' :
                      trend.direction === 'increasing' ? '\u2191' : '\u2194';
    console.log(`\n  Trend: ${trendIcon} ${trend.direction || 'stable'} (${(trend.percent_change || 0).toFixed(1)}%)`);

    // Clusters
    console.log('\n' + '-'.repeat(80));
    console.log(' PATTERN RECOGNITION');
    console.log('-'.repeat(80));

    console.log(`\n  Clusters: ${data.clusters.clusters.length}`);
    console.log(`  Silhouette Score: ${(data.clusters.silhouetteScore || 0).toFixed(3)}`);

    for (const cluster of data.clusters.clusters) {
        console.log(`\n    Cluster ${cluster.cluster_id}: ${cluster.device_count} devices`);
    }

    if (data.clusters.outliers.length > 0) {
        console.log(`\n  Outliers: ${data.clusters.outliers.length} devices`);
    }

    // Insights
    console.log('\n' + '-'.repeat(80));
    console.log(' AI INSIGHTS');
    console.log('-'.repeat(80));

    console.log(`\n  Total Insights: ${data.insights.length}`);

    const criticalInsights = data.insights.filter(i => i.severity === 'critical');
    if (criticalInsights.length > 0) {
        console.log('\n  Critical Insights:');
        for (const i of criticalInsights.slice(0, 3)) {
            console.log(`    \u2757 ${i.title}`);
        }
    }

    const actionable = data.insights.filter(i => i.actionable);
    if (actionable.length > 0) {
        console.log(`\n  Actionable Recommendations (${actionable.length}):`);
        for (const i of actionable.slice(0, 3)) {
            console.log(`    - ${i.title}`);
        }
    }

    // Connectivity
    console.log('\n' + '-'.repeat(80));
    console.log(' CONNECTIVITY');
    console.log('-'.repeat(80));

    const connStatus = data.connectivity.status;
    console.log(`\n  Devices: ${connStatus.online || 0}/${connStatus.total || 0} online`);

    const latencyStats = data.connectivity.latency;
    console.log('\n  Latency:');
    console.log(`    Average: ${(latencyStats.avg_latency_ms || 0).toFixed(1)} ms`);
    console.log(`    P95: ${(latencyStats.p95_latency_ms || 0).toFixed(1)} ms`);

    const throughputStats = data.connectivity.throughput;
    console.log('\n  Throughput (24h):');
    console.log(`    Total Messages: ${(throughputStats.total_messages || 0).toLocaleString()}`);
    console.log(`    Avg/Hour: ${(throughputStats.avg_per_hour || 0).toLocaleString()}`);

    const qualityStats = data.connectivity.quality;
    console.log(`\n  Connection Quality: ${(qualityStats.average_quality_score || 0).toFixed(0)}/100`);

    // Footer
    console.log('\n' + '='.repeat(80));
    console.log(' '.repeat(30) + 'END OF REPORT');
    console.log('='.repeat(80));
}

/**
 * Dashboard refresh loop
 */
async function dashboardRefreshLoop(client, intervalSeconds = 60, maxIterations = 5) {
    console.log('Starting dashboard refresh loop...');
    console.log(`Refresh interval: ${intervalSeconds} seconds`);
    console.log('Press Ctrl+C to stop\n');

    for (let i = 0; i < maxIterations; i++) {
        try {
            const data = await collectDashboardData(client);
            console.log('\n'.repeat(2));
            renderDashboard(data);
            console.log(`\n[Refresh ${i + 1}/${maxIterations} - Next refresh in ${intervalSeconds} seconds]`);
            await new Promise(resolve => setTimeout(resolve, intervalSeconds * 1000));
        } catch (error) {
            console.error(`Error refreshing dashboard: ${error.message}`);
            await new Promise(resolve => setTimeout(resolve, 10000));
        }
    }

    console.log('\nDashboard loop completed.');
}

/**
 * Export dashboard data to JSON
 */
async function exportDashboardJson(client, outputFile = 'dashboard_data.json') {
    console.log(`Exporting dashboard data to ${outputFile}...`);
    const data = await collectDashboardData(client);

    const fs = await import('fs');
    fs.writeFileSync(outputFile, JSON.stringify(data, null, 2));

    console.log(`Dashboard data exported to ${outputFile}`);
}

// Main entry point
async function main() {
    const client = new AnalyticsClient();
    const args = process.argv.slice(2);

    if (args.includes('--loop')) {
        const interval = parseInt(args[args.indexOf('--loop') + 1]) || 60;
        await dashboardRefreshLoop(client, interval);
    } else if (args.includes('--export')) {
        const output = args[args.indexOf('--export') + 1] || 'dashboard_data.json';
        await exportDashboardJson(client, output);
    } else if (args.includes('--help')) {
        console.log('TESAIoT Analytics Dashboard');
        console.log('\nUsage:');
        console.log('  node example-dashboard.js          # Single refresh');
        console.log('  node example-dashboard.js --loop   # Continuous refresh (60s)');
        console.log('  node example-dashboard.js --loop 30  # Custom interval');
        console.log('  node example-dashboard.js --export  # Export to JSON');
    } else {
        console.log('Collecting dashboard data...');
        const data = await collectDashboardData(client);
        renderDashboard(data);
    }
}

main().catch(console.error);
